source/agora/node/EnrollmentManager.d

```diff
@@ -0,0 +1,138 @@
+import agora.common.crypto.Schnorr : Pair;
+
+import std.random;
+import std.stdio;
+
+/*******************************************************************************
+    Defines data for enrollment process
+*******************************************************************************/
+
+public struct EnrollmentData{
+    /// K: UTXO key, A public key matching a frozen UTXO
+    public PublicKey K;
+
+    /// X: random seed, The nth image of random value
+    public Hash X;
+
+    /// n: cycle length, A number within bounds [0; 1008]
+    /// original spec: [0; (Balance(K) % 40,000) * freezing period * 2]
+    public ulong n;
+
+    /// R: signature noise, The initial nonce used for signing
+    public Point R;
+
+    /// S: A signature for the messge H(K, X, n, R) and the key K, using R
+    public Signature S;
+
+    /***************************************************************************
+        Implements hashing support
+        Params:
+            dg = hashing function
+    ***************************************************************************/
+    public void computeHash (scope HashDg dg) const nothrow @safe @nogc
+    {
+        hashPart(this.K, dg);
+        hashPart(this.X, dg);
+        hashPart(this.n, dg);
+        hashPart(this.R, dg);
+    }
+}
+
+/*******************************************************************************
+    Handles data and process for enrollment
+*******************************************************************************/
+
+public struct EnrollmentManager
+{
+    /// Node's key pair
+    private KeyPair node_keypair;
+
+    /// Random key for enrollment
+    private Pair random_key;
+
+    /// Node balance
+    private Amount balance;
+
+    // Random object
+    private Random genint;
+
+    /// EnrollmentData
+    private EnrollmentData* data;
+
+    /***************************************************************************
+        Constructor
+        Params:
+            keypair = the keypair of the owner node
+            balance = account balance of the owner node
+    ***************************************************************************/
+    public this(KeyPair keypair, Amount balance)
+    {
+        this.node_keypair = keypair;
+        this.balance = balance;
+        this.genint = Random(123);
+        this.data = null;
+    }
+
+    /***************************************************************************
+        Makes a data object for enrollment process
+        Returns:
+            a EnrollmentData object
+    ***************************************************************************/
+    public EnrollmentData getEnrollmentData()
+    {
+        this.data = new EnrollmentData();
+        
+        // public key
+        this.data.K = this.node_keypair.address;
+
+        // cycle length
+        this.data.n = uniform(0, 1008, this.genint);
+
+        // random seed
+        this.data.X = hashFull(this.node_keypair.secret.data);
```

Comment:
This is the part that should use the random number generator.

Reply:
The random seed should not be based on any existing data, it should ideally just be a random value. For example:

```D
Hash X = Scalar.random().hashFull();
```

The initial X should be stored somewhere locally so the Nth preimage can be generated from it on demand when a node needs to reveal that Nth preimage later. Either that, or you can just store an array of preimages for quick lookup later.

Reply:
any update on this?

Reply:
Ah, Mathias commented that I could use the constant value, or 1008 at the moment before he left.

Reply:
I'm talking about the random seed value X, not N. We should use a random number generator, and not a hash of the private data.

Reply:
Just to update everyone on this: Initially we planned on using the private key as the seed, which would have to be revealed at the end of the validation cycle. However, after @Geod24 had reservations (even though this part is still in the yellowpaper), so for now we can just make this a random number.

The reason why we need this X, is for the verifiable random number generation in the blockchain. This is important for many things: quorum balancing (https://github.com/bpfkorea/agora/issues/240), contracts (think games, lottery apps), etc.